### PR TITLE
Add AST_CLASS_NAME in AST version 70

### LINKED ISSUE #109: Internal AST representation of `Foo::class` changed in PHP 7.4-dev
Observed: An invalid node is produced when ast\parse_code() is called on code such as `Token::class`
Expected: This produces AST_CLASS_CONST nodes with correct flags on the 'class' child node

```php
<?php declare(strict_types=1);

// This fails in php 7.4
$code = <<<'EOT'
<?php
namespace Foo;
Token::class;  // this has node kind 276 and 'children' => ['Token']
\Token::class;  // this also has node kind 276 and 'children' => ['Token']
Token::other_constant;  // this has node kind 516 and children 'class' and 'constant'
EOT;
$node = \ast\parse_code($code, 50);
var_export($node);
var_export(ast\get_kind_name($node->children[1]->kind)); // throws  Uncaught LogicException: Unknown kind 276 
```

### PR DESCRIPTION
**Pull request by TysonAndre**
Fixes #109

This also makes PHP 7.4 emulate the old behavior (AST_CLASS_CONST) for versions 50 and 60.

Tests pass with a local PHP 7.4 installation

### CODE REVIEW

**Comment by TysonAndre** [comment]
The merge conflicts were fixed, tests in php 7.4-dev are still passing locally

File: ast.c [hunk 1]
```diff
@@ -724,6 +720,50 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
...
+				zend_ast *const_name_ast = ast->child[1];
+				if (const_name_ast->kind != ZEND_AST_ZVAL) {
+					break; // Currently impossible
+				}
+				zval const_name_zv = *zend_ast_get_zval(const_name_ast);
+				if (Z_TYPE(const_name_zv) != IS_STRING) {
+					break; // Currently impossible
+				}
+				zend_string *str = Z_STR(const_name_zv);
+				if (zend_string_equals(AST_STR(str_class), str)) {
```
Comment (nikic) [inline ast.c]: This should be `equals_ci`. `::CLASS` is case-insensitive.

File: ast.c [hunk 2]
```diff
@@ -724,6 +720,50 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
...
 			break;
+		case ZEND_AST_CLASS_NAME:
+			if (state->version < 70) {
+				zval name_zval;
+				ast_to_zval(&name_zval, ast->child[0], state);
+				zval class_name_zval;
+				ast_create_virtual_node_ex(
+						&class_name_zval, AST_NAME, ast->child[0]->attr, zend_ast_get_lineno(ast), state, 1, &name_zval);
+				zval const_zval;
+				ZVAL_STR(&const_zval, AST_STR(str_class));
```
Comment (nikic) [inline ast.c]: ZVAL_STR_COPY, the string may be non-interned in zts builds of older PHP versions.

File: php_ast.h [hunk 3]
```diff
@@ -51,6 +51,7 @@ extern ast_str_globals str_globals;
 
 // 544 is already taken by ZEND_AST_GROUP_USE
 #if PHP_VERSION_ID < 70400
+# define ZEND_AST_CLASS_NAME 0x2fe
```
Comment (nikic) [inline php_ast.h]: This should be `0x1ff`, as it has only one child node.

File: ast.c [hunk 4]
```diff
@@ -724,6 +720,50 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
...
+		case ZEND_AST_CLASS_CONST:
+			if (state->version >= 70) {
+				zend_ast *class_name_ast = ast->child[0];
+				if (class_name_ast->kind != ZEND_AST_ZVAL) {
+					break;
+				}
+				// Convert to an AST_CLASS_NAME instead. The opposite of the work done in the ZEND_AST_CLASS_NAME case.
+				zend_ast *const_name_ast = ast->child[1];
+				if (const_name_ast->kind != ZEND_AST_ZVAL) {
+					break; // Currently impossible
```
Comment (nikic) [inline ast.c]: If it's impossible, we should drop the check. `zend_ast_get_zval` contains an assertion that checks this. Even better would be to use `zend_ast_get_str()` to directly get at the zend_string.

Comment (TysonAndre) [reply ast.c]:
```
php > var_export([]::class);

Fatal error: Cannot use ::class with dynamic class name in php shell code on line 1
```

Okay, it's possible (I just didn't test every possible edge case) - that's a runtime error, not a syntax error

Comment (TysonAndre) [reply ast.c]: Okay, I'm going to have to report a bug with php 7.4-dev debug ZTS on bugs.php.net:

```php
php > var_export((2)::class);                                                                                                                                                                  
php: /path/to/php-src/Zend/zend_ast.h:298: zend_ast_get_str: Assertion `zval_get_type(&(*(zv))) == 6' failed.
[1]    4949 abort      php -a
```

Comment (TysonAndre) [reply ast.c]: Filed https://bugs.php.net/bug.php?id=77530&thanks=4

Comment (TysonAndre) [reply ast.c]: It turns out that I was mistaken and it indeed something I should check for (I tried `2::class` the first time but not `(2)::class`)

**Comment by TysonAndre** [comment]
All review comments have been addressed, this is ready for another look. The new tests continue to pass in PHP 7.4.

Note that the flags on the AST_ARRAY of `[]::class` are inconsistent, but that's very low priority

File: ast.c [hunk 5]
```diff
@@ -724,6 +720,53 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
...
+						zv, ZEND_AST_CLASS_CONST, 0, zend_ast_get_lineno(ast), state, 2, &class_name_zval, &const_zval);
+				return;
+			}
+			break;
+#else
+		case ZEND_AST_CLASS_CONST:
+			if (state->version >= 70) {
+				// Convert to an AST_CLASS_NAME instead. This is the opposite of the work done in the ZEND_AST_CLASS_NAME case.
+				zend_ast *const_name_ast = ast->child[1];
+				if (const_name_ast->kind != ZEND_AST_ZVAL) {
```
Comment (nikic) [inline ast.c]: While the class can be an expression, the constant name can't be (we don't have `Foo::{$bar}` in PHP), so these checks at least can be replaced with `zend_ast_get_str`.

File: ast.c [hunk 6]
```diff
@@ -724,6 +720,53 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
...
+					if (class_name_ast->kind == ZEND_AST_ZVAL) {
+						// e.g. Foo::class
+						zval class_name_raw_zval;
+						ZVAL_COPY(&class_name_raw_zval, zend_ast_get_zval(class_name_ast));
+						ast_create_virtual_node_ex(
+								&class_name_zval, AST_NAME, class_name_ast->attr, zend_ast_get_lineno(class_name_ast), state, 1, &class_name_raw_zval);
+					} else {
+						// e.g. []::class (not a parse error, but a runtime error)
+						ast_to_zval(&class_name_zval, class_name_ast, state);
+					}
```
Comment (nikic) [inline ast.c]: Wouldn't `ast_to_zval` here work for both branches?

Comment (TysonAndre) [reply ast.c]: It didn't create the `ast\Node` when I tried that, probably because the parent kind or position was wrong.

Comment (nikic) [reply ast.c]: Ah yeah, I guess it will just convert it to a simple value. In that case it's fine.

**Comment by nikic** [comment]
Thanks!
